File: src/scheme/evaluator/call-cc.ts
import { type Thunk, type SchemeData, SchemeCont, SchemeList, SchemeProc, SchemeSym } from '../parser/data'
import { Env, StackFrame } from '../env'
import type { IEvaluator, Evaluator } from './index'
import { assert } from '../utils'

/**
 * 语法（后面加一个 lambda 语句）：
 * (call-with-current-continuation
 *    (lambda (resume-here)
 *      (set! control-state resume-here)
 *      (return element)))
 */
export default class CallCCEvaluator implements IEvaluator {
  constructor(private evaluator: Evaluator) {}

  public matches(node: SchemeData): boolean {
    return SchemeSym.matches(node) && node.value === 'call-with-current-continuation'
  }

  public evaluate(node: SchemeList, env: Env, cont: SchemeCont): Thunk {
    return this.evaluator.evaluate(node.cadr(), env, new SchemeCont((proc: SchemeData) => {
      assert(SchemeProc.matches(proc), 'callcc args evaluate eror!')
      const newEnv = new Env(env, new StackFrame(env.getStackFrame()))
      const virtualName = 'callcc'
      const virtualNode = SchemeList.buildFromArray([
        new SchemeSym(virtualName),
        this.buildEscapedProc(env, cont)
      ])
      newEnv.define(virtualName, proc)
      return this.evaluator.evaluate(virtualNode, newEnv, cont)
    }))
  }

  private buildEscapedProc(env: Env, cont: SchemeCont): SchemeProc {
    const paramName = '(escaped-proc-param)'
    const params = SchemeList.buildFromAtom(new SchemeSym(paramName))
    const virtualNode = SchemeList.buildFromArray([cont as any, new SchemeSym(paramName)])
    const body = SchemeList.buildFromAtom(virtualNode) // need to wrap by a list
    return new SchemeProc('<<captured contiuation>>', params, body, env)
  }
}

File: packages/interpreter-rust-new/src/evaluator/call_cc.rs
use std::collections::VecDeque;

use crate::{parser::{SchemeExp, SchemeData, SchemeCont}, env::Env};

use super::Evaluator;

/**
 * 语法（后面加一个 lambda 语句）(**此表达式有返回值**)：
 * (call-with-current-continuation
 *    (lambda (resume-here)
 *      (set! control-state resume-here)
 *      (return element)))
 */
impl Evaluator {
  pub fn evaluate_call_cc(&mut self, mut node: SchemeExp, env: Env, next: usize) -> usize {
    node.value.pop_front();

    let lambda = node.value.pop_front().expect("Evaluate callcc-lambda error!");
    if let SchemeData::Exp(x) = lambda {
      let mut proc = self.evaluate_lambda_to_proc(x, env.clone());
      let cont = SchemeData::Continuation(SchemeCont {
        value: next,
        loc: None
      });
      return self.evaluate_proc_with_arguments(&mut proc, &mut VecDeque::from([cont]), env.copy(), next)
    }

    panic!("Evaluate callcc error!")
  }
}